File: 0020 - Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        l = []
        for p in s:
            if p in ['{', '(', '[']:
                l.append(p)
            else:
                if l and ((p == '}' and l[-1] == '{') or (p == ']' and l[-1] == '[') or (p == ')' and l[-1] == '(')):
                    l.pop()
                else:
                    return False
        
        if len(l):
            return False
        return True

File: 0020 - Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        vector<char> anticlockwise = {'(', '[', '{'};
        vector<char> clockwise = {')', ']', '}'};
        
        stack<int> parentheses;

        for(auto& character: s){
            if(character )
            if(find(anticlockwise.begin(), anticlockwise.end(), character) != anticlockwise.end()){
                parentheses.push(character);
            }
            auto it = find(clockwise.begin(), clockwise.end(), character);
            if(it != clockwise.end()){
                int index = it - clockwise.begin();
                if(parentheses.empty() || parentheses.top() != anticlockwise[index]) return false;
                parentheses.pop();
             }
        }
        return parentheses.empty();
    }
};